### PR TITLE
DEV-2768: Teach skills and agent docs the determinism rules the flake work earned

### PR DESCRIPTION
**Pull request by danielzytohoc**
### Context

The PR adds the determinism rules the CI-flakiness investigation earned to the skills and agent docs that dispatch test work, so the next migration inherits them instead of re-learning them run by run. Skills, agent docs, `AGENTS.md`, and `.ai/` files, plus one `node --test` check over the review dispatch surfaces; no product code, no changelog entry (tooling/test-infrastructure change).

The text describes the tree after the three sibling lint PRs land, and they merge first: #13349 (the global `setTimeout` at error in specs and page objects, frame waits flagged in the frozen tier), #13350 (the diff-scoped ratchet), and #13364 (every `waitForFunction()` passes `{ polling }`, and `tests/.eslintrc.cjs` errors on one without it). This PR merges develop before landing. Where it overlaps #13364 in `tests/AGENTS.md` (the fixture-contract `goto()` bullet), the wording is byte-identical to that branch.

Each rule lands where the dispatch already points readers, with the one-line WHY drawn from the measured incident:

- **Playwright mechanics** (`handsontable-playwright-e2e` + `references/determinism.md`, `tests/AGENTS.md`): the six page-object rules lint cannot see — lint reads source text, so a timer in a fixture's inline script or a string-form `evaluate`, and the state a wait ends on, stay out of its reach. A `setTimeout` that runs in the browser is a fixed wait wherever it sits; `page.waitForFunction()` passes an explicit `{ polling }` (the rAF default starved under worker load: 3 mute timeouts in ~700 runs of `hidden-init-rerender.spec.ts`, 0 after timer polling — every page object passes it since #13364); a scroll/mutate method ends on a render-state probe, never `scrollTop` (reference `OverlaysPage.scrollToEnd()`, which ends on the last cell being rendered; `FrozenTallCellPage.scrollVerticallyTo()` is the counter-example behind the `frozen-column-row-heights` gap); a multi-delivery trigger is asserted by polling the latest entry of its kind; a fixture build fails loud through a captured constructor throw; a negative assertion needs a positive control. `references/determinism.md` carries the full rules and the incident behind each; `tests/AGENTS.md` carries a one-sentence summary per rule and points there, so there is one copy of the prose to keep current. Its fixture-contract bullet spells out `{ polling: 100 }`.
- **Judgment** (`test-writing-discipline`): a weakened/deleted assertion over an admitted race ships with a ticket in the same PR; a timing-semantics JSDoc claim is checked against its primitive (`setTimeout(0)` is the next macrotask, not end-of-task — that mismatch shipped #13332); an X-during-Y implementation path needs a test that drives X during Y; every documented option form executes once; "passes on retry / in isolation" is not determinism evidence — hammer it with `--repeat-each` across all legs, and a ~50% failure on unchanged develop is a product ticket.
- **Frozen tier** (`handsontable-e2e-testing`, `handsontable/.ai/TESTING.md`): rendered-DOM count assertions pin the viewport first; `waitUntil(condition, timeout)` replaces `sleep()` / `waitForNextAnimationFrames()` (a fixed sleep denominated in frames, capped at 2 real frames); the flake-ledger trigger — red/flaky across ≥2 distinct CI runs → migration ticket next sprint; new `it.flaky()` is a migration signal. The `sleep()` wording matches the lint: `handsontable/no-fixed-sleep-in-spec` warns on every call (warn level so the existing debt surfaces without red-walling CI), and the diff-scoped ratchet (#13350) fails a NEW `sleep()` on an added line. The code sample shows `waitUntil()` and marks `sleep()` and `waitForNextAnimationFrames()` as never, so the sample and the helper list say the same thing.
- **Review lens** (`handsontable-code-review`, new `references/tests.md`): a fourth "Tests" dimension — name the exercising test per new interaction path, run scoped mutation when unit tests changed (the one run a review makes; in a `.claude/worktrees/` checkout it dies with `env-cmd: command not found` until `node scripts/claude/setup-worktree.mjs` runs, and the reference says so), flag near-duplicate DOM helpers, check timing JSDoc against the primitive, confirm the ticket on a weakened assertion. The lens reaches every surface that dispatches a review: the skill's frontmatter description, the `handsontable-reviewer` agent (four dimensions, the `tests.md` bullet, the mutation-run exception, `/ Tests` in the output grouping), and both Bugbot context files (`.cursor/BUGBOT.md`, `handsontable/.cursor/BUGBOT.md`). `scripts/__tests__/review-dispatch.test.mjs` pins that: every `references/*.md` the skill ships must be cited by a path a reader can open — the repo path from the three surfaces outside the skill, the skill-relative `references/<name>` from `SKILL.md` only — with a positive control on the directory read so an empty listing cannot pass vacuously. Another skill's `references/` folder does not count (two skills ship one). A second check pins the reviewer agent's one-line JSDoc summary to `references/code-quality.md` on the `[[Target]]`/`{@link}` pair: the agent still told reviewers `[[LINK]]` after #13339 moved the reference to `{@link}`, so the line is fixed and `code-quality.md` carries develop's post-#13339 text so the branch agrees with itself. It runs under `npm run test:tooling` (CI job `tooling tests`), reads six files, and needs no build.
- **Test-writer agent** (`handsontable-test-writer`): reaches `test-writing-discipline` and `handsontable-playwright-e2e` (it pointed only at the unit, frozen-Jasmine, and walkontable skills), routes new E2E to Playwright with a short conventions block (`failOnFlakyTests` is on in CI, not locally — hammer with `--repeat-each` before calling a spec deterministic), and labels the Jasmine block as frozen/edit-only — an agent following the old text would write a new `*.spec.js` and hit the presence gate.
- **Routing** (`walkontable-dev`, `handsontable-dev`, `walkontable/AGENTS.md`): rendering-engine geometry changes get engine-tier specs — existing walkontable Jasmine specs may be edited, new coverage goes to `tests/e2e/walkontable/*.spec.ts`. `walkontable-dev` names `OverlaysPage.scrollToEnd()` as the scroll-method shape to copy and `FrozenTallCellPage.scrollVerticallyTo()` as the counter-example.

**Merge note (develop merged after the four enforcement PRs landed):** the two conflicts were the paragraphs this branch and those PRs describe from different angles — `tests/AGENTS.md`'s `goto()` snippet now shows `awaitBundle()`, and its Determinism section keeps the widened ban paragraph above this branch's six-rule summary, whose polling bullet points at the helper that owns the interval; `handsontable/.ai/TESTING.md` keeps the fuller `waitUntil()` paragraph with this branch's stronger "never" markers and the ratchet consequence. A first merge also fixed four page objects that `develop` was briefly red on (unpolled bundle waits from branches that predated the polling rule); develop then fixed the same four itself (#13421, #13422), so the second merge takes develop's versions and this PR changes nothing under `tests/fixtures/pages`.

### How has this been tested?

- `scripts/__tests__/review-dispatch.test.mjs` was written first and run against the tree before the dispatch edits: `# pass 3 / # fail 3` — the three failures were `.claude/agents/handsontable-reviewer.md`, `.cursor/BUGBOT.md`, and `handsontable/.cursor/BUGBOT.md`, each reporting exactly `references/tests.md` as the missing citation (the skill's own `SKILL.md` and the helper's self-check passed). After the edits: `# pass 6 / # fail 0`.
- Review round: the citation matcher now requires the full path outside the skill. The old bare-substring matcher returned `[]` for a surface citing `.claude/skills/handsontable-playwright-e2e/references/tests.md`; the new one reports `tests.md` missing, and a unit test pins that input. The JSDoc link-form check fails on the agent's stale `[[LINK]]`, not `{@link}` line against the post-#13339 reference (a unit test pins the literal old lines). After the fixes: `# tests 10 / # pass 10 / # fail 0`.
- `npm run test:tooling` — `# tests 372 / # pass 372 / # fail 0` (the whole enforcement-machinery suite, new file included).
- `npm run eslint` (root config over `scripts/` and `.github/scripts/`) — clean.
- `node scripts/claude/setup-worktree.mjs` — worktree bootstrapped (exit 0).
- `node scripts/sync-skills-to-cursor.mjs` — `Done. Cursor rules are in sync with Claude skills.` (exit 0; `.cursor/rules/` is gitignored).
- Added lines scanned for `quality/` references (none — `.ai/LOCAL-ENFORCEMENT.md` §3), for the banned evaluative adjectives and British spellings from `.ai/DOC-STANDARDS.md` (none), and for task IDs other than this PR's (none).
- The repo lints no Markdown (no markdownlint/remark config). Pre-push `test-gate`: a docs-plus-tooling-test change is correctly out of the presence gate's scope.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):

1. DEV-2768

### Affected project(s):

- [ ] `handsontable`
- [ ] `@handsontable/angular-wrapper`
- [ ] `@handsontable/react-wrapper`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://cla.handsontable.com/sign) — one signature covers both Handsontable and HyperFormula; the `cla/signed` check on this PR confirms it.
- [ ] My change requires a change to the documentation.
- [ ] This change needs a manual QA pass

ClickUp task: https://app.clickup.com/t/9015210959/DEV-2768

---

> [!NOTE]
> **Low Risk**
> Documentation and tooling-test changes only; no runtime or public API behavior is modified.
> 
> **Overview**
> Documents flake-investigation learnings across agent skills, `AGENTS.md`, and `.ai/TESTING.md` so reviewers and test writers inherit the same rules without re-discovering them per migration.
> 
> **Review** gains a fourth **Tests** dimension via new `references/tests.md` (named exercising tests per path, scoped mutation on unit-test diffs, timing JSDoc vs primitives, tickets on weakened assertions). That checklist is wired through `handsontable-code-review`, the `handsontable-reviewer` agent, and both Bugbot contexts; the reviewer agent’s JSDoc one-liner is aligned with `code-quality.md` (`{@link}` not `[[Target]]`).
> 
> **Authoring** updates steer new E2E to Playwright (`handsontable-test-writer`, expanded `determinism.md` with six page-object wait rules lint cannot see), tighten frozen Jasmine edits (`waitUntil`, viewport pinning, flake ledger), and extend `test-writing-discipline` (same-PR tickets for race-driven weakening, `--repeat-each` before calling specs deterministic).
> 
> **Routing** clarifies Walkontable geometry specs → `tests/e2e/walkontable/*.spec.ts`. **`scripts/__tests__/review-dispatch.test.mjs`** asserts every review dispatch surface cites each `handsontable-code-review/references/*.md` by an openable path (and rejects wrong-skill `references/` matches).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8fe8182e1a8b2fc6c48a0787fb39ab08510fb062. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [DEV-2768 Skills and agent docs: the determinism rules the flake investigation earned (polling intervals, render-state probes, latest-of-kind polls, ticket-on-weakening, feature-interaction and option-forms)](https://app.clickup.com/t/9015210959/DEV-2768)

File: .claude/skills/handsontable-playwright-e2e/references/determinism.md [hunk 1]
```diff
@@ -7,3 +7,14 @@ A flake-free Playwright spec:
 - Freeze time with `page.clock` where behavior depends on it; mock network with `page.route`.
 - Keep fixtures small — a big dataset slows every test and makes flakes likelier.
 - `failOnFlakyTests` is on in CI: a test that only passes on retry is a hard failure. Fix the root cause, don't add retries.
+
+## Waits inside a page object
+
+Lint bans the fixed waits it can see in a spec. These six live one level down — inside `page.evaluate()`, a `waitForFunction`, or a probe method — where lint cannot follow, so the page object has to get them right. Each one was measured on a real migration.
```
Comment (budnix) [inline .claude/skills/handsontable-playwright-e2e/references/determinism.md]: PR #13349 bans `setTimeout` in this tier at error level, and its own description says the point is catching a timer inside `page.evaluate()` — which is item 1 right below this line. It also rewrites line 5 of this same file, and edits `tests/AGENTS.md` and `handsontable/.ai/TESTING.md`, so the two PRs touch the same three files. Whichever one merges second will need this sentence reworded, and git may auto-merge them cleanly and leave both claims sitting here. Worth deciding now which of the two lands first.

File: .claude/skills/handsontable-playwright-e2e/references/determinism.md [hunk 2]
```diff
@@ -7,3 +7,14 @@ A flake-free Playwright spec:
...
+
+1. **`setTimeout` inside `page.evaluate()` is a fixed wait.** It is `sleep()` moved into the browser, and the ban's reason moves with it. Expose the state as a data probe method and `expect.poll` it from the spec.
+2. **`page.waitForFunction()` passes an explicit `{ polling: <ms> }`.** The default polls on `requestAnimationFrame`, and parallel-worker load can starve rAF callbacks past the timeout while the page is healthy — 3 mute timeouts in ~700 runs of `hidden-init-rerender.spec.ts`, 0 after switching to timer polling. Reference: `HiddenInitRerenderPage.goto()`. The page objects written before this rule still open with `waitForFunction(() => 'Handsontable' in window)` on the default — known debt, not a second pattern to copy; a sweep that moves them to timer polling is ticketed. A new page object follows the reference, whatever its neighbors do.
```
Comment (budnix) [inline .claude/skills/handsontable-playwright-e2e/references/determinism.md]: This says the sweep is ticketed but does not say which ticket, and `tests/AGENTS.md` line 179 has the same gap. It looks like DEV-2769 / PR #13364, which is open right now. This PR does name #13332, #13334 and #13336 over in `TESTING.md` and `test-writing-discipline`, and the rule it adds in `references/tests.md` says debt without a named ticket blocks a review — so naming it here too keeps the file honest by its own standard.

File: .claude/skills/walkontable-dev/SKILL.md [hunk 3]
```diff
@@ -114,6 +114,7 @@ Walkontable has its own dedicated test runner. Do NOT mix Walkontable tests with
 - **Run tests**: `npm run test:walkontable --prefix handsontable`
 - **Test location**: `src/3rdparty/walkontable/test/`
 - Always test with frozen rows and columns enabled to cover overlay edge cases.
+- **Where a geometry change gets its spec.** A change to viewport calculation, overlay positioning, row or column sizing, or scroll sync gets an engine-tier spec, not a core-tier one. The existing Jasmine specs under `src/3rdparty/walkontable/test/` may be edited; new coverage goes to `tests/e2e/walkontable/*.spec.ts` with a page object in `tests/fixtures/pages/walkontable/` (reference: `frozen-column-row-heights.spec.ts` + `FrozenTallCellPage.ts`). A page-object method that scrolls ends on a render-state probe (the first rendered row, a draw counter), never on `scrollTop` — the redraw is rAF-batched and lands after the scroll position settles. Rules: `handsontable-playwright-e2e`, `references/determinism.md`.
```
Comment (budnix) [inline .claude/skills/walkontable-dev/SKILL.md]: This says a scrolling page-object method must end on a render-state probe, then points at `FrozenTallCellPage.ts` as the example to copy. But `scrollVerticallyTo()` in that file ends on `expect.poll(... el.scrollTop)`, and `determinism.md` line 17 calls that exact method "the frozen-column-row-heights gap". So anyone following this row copies the shape the same sentence forbids. Could it point at a page object that already follows the rule, or say outright that this one is the counter-example?

File: tests/AGENTS.md [hunk 4]
```diff
@@ -164,3 +166,30 @@ naming the task (`// eslint-disable-next-line no-restricted-syntax --
 DEV-1234: <why>`), which keeps every parked test counted and attributable.
 Full rules: the `handsontable-playwright-e2e` skill and its
 `references/determinism.md`.
+
+The waits lint cannot see live in page objects. Never get these wrong:
```
Comment (budnix) [inline tests/AGENTS.md]: All six rules are repeated word for word here, two lines under the pointer that says the full rules live in `references/determinism.md`. That leaves two copies to keep in step, and a later fix to rule 2 or 3 in one file will leave the other quietly stale. This PR ships `review-dispatch.test.mjs` to stop exactly that drift for the review surfaces, so it feels odd to leave the largest copy unguarded. Would a short summary plus the pointer work here instead?

File: handsontable/.ai/TESTING.md [hunk 5]
```diff
@@ -275,9 +275,10 @@ Do **not** add `await sleep(100)` / `sleep(200)` to new tests. A fixed delay is
 ```javascript
 // legacy Jasmine E2E: await the state, not the clock
 await selectCell(0, 0);                 // helpers already await the scroll they trigger
+await waitUntil(() => spy.calls.count() > 0);  // the condition-based waiter (test/helpers/common.js)
 await sleep(...);                        // ← avoid; if you must, root-cause it and add a real waiter
 ```
-For new Playwright tests use web-first, auto-retrying assertions (`await expect(locator).toBeVisible()`); see the `handsontable-playwright-e2e` skill. Existing `sleep()` sites are baselined by lint, but a broken or flaky legacy test is a signal to migrate it (see below), not to add another delay.
+For new Playwright tests use web-first, auto-retrying assertions (`await expect(locator).toBeVisible()`); see the `handsontable-playwright-e2e` skill. Every `sleep()` call warns today — `handsontable/no-fixed-sleep-in-spec` runs at warn level so the existing debt surfaces without red-walling CI — and a diff-scoped ratchet that fails a NEW `sleep()` on an added line is landing in a sibling PR. A broken or flaky legacy test is a signal to migrate it (see below), not to add another delay.
```
Comment (budnix) [inline handsontable/.ai/TESTING.md]: The sibling PR here is DEV-2765 / PR #13350, which is open now — worth naming it. As written a reader cannot tell whether the ratchet actually shipped, and if that PR gets closed instead of merged this line quietly becomes wrong. Line 646 makes the same claim, so both spots would need the id.

File: handsontable/.ai/TESTING.md [hunk 6]
```diff
@@ -639,7 +642,8 @@ it('should maintain selections after render', async() => {
 - `createSpreadsheetData(rows, cols)` -- Grid with coordinates as values ('A1', 'B2', etc.)
 
 **Async Utilities:**
-- `sleep(delay = 100)` -- Promise-based delay
+- `waitUntil(condition, timeout = 4000)` -- Polls `condition` every frame; rejects with a named reason on timeout. The replacement for `sleep()` and `waitForNextAnimationFrames()` in an edited spec
+- `sleep(delay = 100)` -- Promise-based delay (every call lint-warns today, and a diff-scoped ratchet that fails a new one on an added line is landing; never add one)
```
Comment (budnix) [inline handsontable/.ai/TESTING.md]: This says never add a `sleep()`, but the code sample at line 279 still reads "avoid; if you must, root-cause it and add a real waiter". That is two different strengths of the same rule in one file, and the sample is the part people copy. Could the sample show `waitUntil()` instead, so both spots say the same thing?

File: .claude/agents/handsontable-test-writer.md [hunk 7]
```diff
@@ -31,7 +31,13 @@ The authoritative, step-by-step conventions live in the testing skills — read
 - Module aliases: `'handsontable'` -> `src/`, `'walkontable'` -> `src/3rdparty/walkontable/src/`
 - Run: `npm run test:unit --prefix handsontable --testPathPattern=<path>`
 
-## E2E test conventions (Jasmine/Puppeteer)
+## E2E test conventions (Playwright — all new E2E)
+
+- File: `tests/e2e/*.spec.ts`; selectors and flows in a page object under `tests/fixtures/pages/`; import `test` from `tests/fixtures/test.ts` and pass `{ theme, bundle }` to the page object
+- Web-first waits only (`await expect(locator)…`, `expect.poll`) — never `sleep`/`waitForTimeout`; a page object's `waitForFunction` passes `{ polling }`, and a scroll method ends on a render-state probe (`handsontable-playwright-e2e/references/determinism.md`)
+- Run only the spec you changed: `cd tests && npx playwright test e2e/<spec>.spec.ts` (all theme × bundle legs; `failOnFlakyTests` is on)
```
Comment (budnix) [inline .claude/agents/handsontable-test-writer.md]: `playwright.config.ts` sets `retries: process.env.CI ? 1 : 0` and `failOnFlakyTests: !!process.env.CI`, so on the local command given here nothing retries and no flaky verdict can ever be reported. A test author runs this, sees green, and reads it as proof the spec is not flaky. `handsontable-playwright-e2e/SKILL.md` words it as "on in CI", which is the accurate version.
```suggestion
- Run only the spec you changed: `cd tests && npx playwright test e2e/<spec>.spec.ts` (all theme × bundle legs; `failOnFlakyTests` is on in CI, not locally — hammer with `--repeat-each` before calling it deterministic)
```

File: .claude/skills/handsontable-code-review/references/tests.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+## Checklist
+
+1. **Every new interaction path has a named exercising test.**
+   - For each new branch that handles one operation landing inside another — `flush`, `drain`, `cancel`, `pending`, `suspend` methods and depth counters are the tell — name the test that drives X *during* Y. A test that runs them in sequence does not count.
+   - For each documented form of a new option, name the test that executes it. The guide's lead form is the one most often left untested.
+   - If you cannot name the test, the finding is **High**: the path shipped unexercised.
+
+2. **Run scoped mutation when unit tests changed.**
+   - This is the one run a review makes. From `handsontable/`: `node ../evals/score.mjs <test.unit.js> --mutate <src.ts>` (`evals/README.md` has the setup). Always scope to the changed source — never the whole tree.
```
Comment (budnix) [inline .claude/skills/handsontable-code-review/references/tests.md:13]: The reviewer agent often runs inside a `.claude/worktrees/` checkout, and the root CLAUDE.md gotcha says those have a symlinked root `node_modules` with the package-local `.bin` missing. `score.mjs` shells `npx env-cmd -f ../hot.config.js npx stryker run` from `handsontable/`, so there it dies with `env-cmd: command not found` and exit 127. A reviewer hitting that has no way to read it as an environment problem, so one line pointing at `node scripts/claude/setup-worktree.mjs` would save them.

File: scripts/__tests__/review-dispatch.test.mjs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * Names the reference files a dispatch surface does not cite by path.
+ *
+ * @param {string} source The dispatch surface's content.
+ * @param {string[]} references The reference file names the skill ships.
+ * @returns {string[]} The reference file names the surface omits.
+ */
+function missingCitations(source, references) {
+  return references.filter(name => !source.includes(`references/${name}`));
```
Comment (budnix) [inline scripts/__tests__/review-dispatch.test.mjs]: This matches the bare substring `references/<name>`, so a surface citing another skill's `references/` folder would pass the check without ever pointing at the code-review one. Two skills in this repo ship a `references/` directory and this PR edits both, so the collision is one shared filename away. The surfaces already spell out the full path, so matching on it closes the hole.
```suggestion
  return references.filter(name => !source.includes(`${SKILL_DIR}/references/${name}`));
```

File: .claude/agents/handsontable-reviewer.md [hunk 10]
```diff
@@ -10,13 +10,14 @@ tools:
...
 - **Code quality** — `.claude/skills/handsontable-code-review/references/code-quality.md`: custom ESLint rules, JSDoc (Markdown links `[[LINK]]`, not `{@link}`; no `<br>`), naming, cognitive complexity (<= 15), silent-catch comments, browser targets, DRY, the TypeScript boundary.
 - **Performance & accessibility** — `.claude/skills/handsontable-code-review/references/performance-a11y.md`: large-array safety, render batching, memory cleanup, WCAG 2.1 AA, both keyboard-navigation modes, ARIA semantics.
+- **Tests** — `.claude/skills/handsontable-code-review/references/tests.md`: a named exercising test for every new interaction path and option form, scoped mutation when unit tests changed, near-duplicate DOM helpers, timing-semantics JSDoc checked against its primitive, a ticket on every weakened or deleted assertion.
```
Comment (budnix) [inline .claude/agents/handsontable-reviewer.md:20]: The new test checks that each reference file is cited, but not that the summary next to it still matches what the reference actually says. The bullet right above is proof: develop's `code-quality.md` line 17 now says use `{@link}` and never TypeDoc `[[Target]]` after #13339, while line 18 here still tells reviewers the opposite. That one did not come from this PR, but it is the exact drift the new test was written to catch, so it may be worth widening the check or just fixing the stale line while you are in this block.

File: .claude/skills/handsontable-playwright-e2e/references/determinism.md (hunk 1 above)
Comment (danielzytohoc) [reply .claude/skills/handsontable-playwright-e2e/references/determinism.md]: Reworded for the tree after #13349 lands: lint (tests/.eslintrc.cjs) now bans the global setTimeout in specs and page objects, including inside a page.evaluate() callback, so the intro no longer claims it cannot see that. What stays out of its reach is stated precisely (a timer inside a fixture HTML page or a string-form evaluate, and the state a wait ends on), and item 1 says the same. Merge order: #13349, #13350 and #13364 land first; this PR then merges develop and resolves the shared hunks in tests/AGENTS.md and TESTING.md there (the determinism.md line 5 change and this section do not overlap, and they say the same thing).

File: .claude/skills/handsontable-playwright-e2e/references/determinism.md (hunk 2 above)
Comment (danielzytohoc) [reply .claude/skills/handsontable-playwright-e2e/references/determinism.md]: Both sentences now state the post-sweep truth instead of an unnamed ticket: every page object passes { polling } since #13364 (checked on that branch: all 13 waitForFunction sites under tests/ carry polling: 100), and tests/.eslintrc.cjs errors on a call without it, with the literal-vs-variable scope spelled out the way #13364's own AGENTS.md text puts it. tests/AGENTS.md got the same replacement in its one-line summary. #13364 merges before this PR, so the wording is true on develop when this lands.

File: .claude/skills/walkontable-dev/SKILL.md (hunk 3 above)
Comment (danielzytohoc) [reply .claude/skills/walkontable-dev/SKILL.md]: Good catch — the row pointed at the counter-example. It now keeps FrozenTallCellPage.ts as the tier/layout reference only, names OverlaysPage.scrollToEnd() as the scroll-method shape to copy (it ends on the last cell being rendered, a render-state probe), and says outright that FrozenTallCellPage.scrollVerticallyTo() ends on scrollTop and is the counter-example whose spec polls masterFirstRenderedRow() itself. determinism.md item 3 carries the same positive reference so the two agree.

File: tests/AGENTS.md (hunk 4 above)
Comment (danielzytohoc) [reply tests/AGENTS.md]: Took the summary option: tests/AGENTS.md now carries one sentence per rule and points at references/determinism.md for the full text and the incident behind each, so the prose lives in one place. The goto() paragraph above stays byte-identical to #13364's wording so the two branches merge cleanly there.

File: handsontable/.ai/TESTING.md (hunk 5 above)
Comment (danielzytohoc) [reply handsontable/.ai/TESTING.md]: Both spots name #13350 now, and the sentences are written for after it merges (it lands before this PR): the ratchet fails a new sleep() on an added line in pre-push and CI, which is exactly where #13350 wires it (lefthook test-gate and the lint workflow step). A reader can open the PR and see whether that is true on develop.

File: handsontable/.ai/TESTING.md (hunk 6 above)
Comment (danielzytohoc) [reply handsontable/.ai/TESTING.md]: The sample now shows waitUntil() as the way to wait (two forms) and marks sleep() and waitForNextAnimationFrames(2) as never, with the reason on each line, so it says what line 648 and the frozen-tier bullet say. The header sentence and the sample lines are #13349's, so inside the code block only the two 'never' comments differ from that branch; the paragraph under it and the async-helper list are shared hunks with #13349 and #13350 that get resolved when this PR merges develop.

File: .claude/agents/handsontable-test-writer.md (hunk 7 above)
Comment (danielzytohoc) [reply .claude/agents/handsontable-test-writer.md]: Accepted as suggested. The line now says failOnFlakyTests is on in CI, not locally, and tells the author to hammer with --repeat-each before calling the spec deterministic — the same wording handsontable-playwright-e2e and test-writing-discipline use.

File: .claude/skills/handsontable-code-review/references/tests.md (hunk 8 above)
Comment (danielzytohoc) [reply .claude/skills/handsontable-code-review/references/tests.md:13]: Added one line under the mutation run: in a .claude/worktrees/ checkout the npx env-cmd step dies with env-cmd: command not found (exit 127) because the package-local .bin is missing behind the symlinked root node_modules — reproduced in a fresh symlinked worktree — that is the environment and not a verdict on the tests, and node scripts/claude/setup-worktree.mjs once fixes it. One detail for the reader: score.mjs itself exits 0 and surfaces that failure as mutation.reason 'stryker run failed: Command failed: … npx env-cmd …', so that reason string is what a reviewer will actually be looking at.

File: scripts/__tests__/review-dispatch.test.mjs (hunk 9 above)
Comment (danielzytohoc) [reply scripts/__tests__/review-dispatch.test.mjs]: Accepted, with one adjustment: SKILL.md is itself a dispatch surface and cites its own references as `references/architecture.md`, so the literal suggestion would have failed it. The matcher now requires the full repo path from every surface outside the skill and accepts the skill-relative form only inside the skill directory and only when nothing path-like precedes it. A new test feeds it `.claude/skills/handsontable-playwright-e2e/references/tests.md` and expects tests.md reported missing from both an outside surface and SKILL.md; the old bare-substring function returned [] for that same input.

File: .claude/agents/handsontable-reviewer.md (hunk 10 above)
Comment (danielzytohoc) [reply .claude/agents/handsontable-reviewer.md:20]: Fixed the stale line: the agent now says {@link Class#member} cross-references, never TypeDoc's [[Target]]. The check is widened for the one pair with two literal forms to name: a test reads which link form references/code-quality.md prescribes (the form it does not negate) and asserts the agent's summary line prescribes the same one, with a unit test pinning the literal old lines. The branch predates #13339, so code-quality.md line 17 carries develop's exact post-#13339 text to make the check hold here; it is byte-identical to develop and merges clean.

**Review by budnix — DISMISSED** [review]

**Review by budnix — DISMISSED** [review]

**Review by budnix — APPROVED** [review]
